content/en/blog/2024/otel-generative-ai/index.md

```diff
@@ -0,0 +1,200 @@
+[Log events API](/docs/specs/otel/logs/event-api/) specification in the Semantic
+Conventions for Generative AI. The events API allows for us to define specific
+[semantic conventions](/docs/specs/semconv/general/events/) for the user prompts
+and model responses that we capture. {{% /alert %}}
+
+### Extending Observability with Vendor-Specific Attributes
+
+The Semantic Conventions also define vendor-specific attributes for platforms
+like OpenAI and Azure Inference API, ensuring telemetry captures both general
+and provider-specific details. This added flexibility supports multi-platform
+monitoring and in-depth insights.
+
+## Building the Python Instrumentation Library for OpenAI
+
+This Python-based library for OpenTelemetry captures key telemetry signals for
+OpenAI models, providing developers with an out-of-the-box observability
+solution tailored to AI workloads. The library,
+[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
+automatically collects telemetry from OpenAI model interactions, including
+request and response metadata and token usage.
+
+As generative AI applications grow, additional instrumentation libraries for
+other languages will follow, extending OpenTelemetry support across more tools
+and environments. The current library's focus on OpenAI highlights its
+popularity and demand within AI development, making it a valuable initial
+implementation.
+
+### Example Usage
+
+Here's an example of using the OpenTelemetry Python library to monitor a
+generative AI application with the OpenAI client.
+
+Install the OpenTelemetry dependencies:
+
+```shell
+pip install opentelemetry-distro
+opentelemetry-bootstrap -a install
+```
+
+Set the following environment variables, updating the endpoint and protocol as
+appropriate:
+
+```shell
+OPENAI_API_KEY=<replace_with_your_openai_api_key>
+
+OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
+OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
+OTEL_SERVICE_NAME=python-opentelemetry-openai
+OTEL_LOGS_EXPORTER=otlp_proto_http
+OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED=true
+# Set to false or remove to disable log events
+OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
+```
+
+Then include the following code in your Python application:
+
+```python
+import os
+from openai import OpenAI
+
+client = OpenAI()
+chat_completion = client.chat.completions.create(
+    model=os.getenv("CHAT_MODEL", "gpt-4o-mini"),
+    messages=[
+        {
+            "role": "user",
+            "content": "Write a short poem on OpenTelemetry.",
+        },
+    ],
+)
+print(chat_completion.choices[0].message.content)
+```
+
+And then run the example using `opentelemetry-instrument`:
+
+```shell
+opentelemetry-instrument python main.py
```

Comment:
Is it expected for this to be exporting to console or otlp? Otlp is used by default but if we want a simple get started for customers perhaps this would be better:

```
opentelemetry-instrument --traces_exporter console --metrics_exporter console python .\server_automatic.py
```

Reply:
btw, if user does NOT set up collector and leaves the example as it is to use otlp by default they will get error messages.

Reply:
I added it as an additional example. We don't have metrics yet, so I just included traces option. Thanks!

Reply:
Nice! If you don't want to include metrics, you have to explicitly set to None.

`opentelemetry-instrument --traces_exporter console --metrics_exporter None python main.py`

Reply:
In that case, I'll just make it console